addons/edi_stock/models/edi_move_cancellation_report_document.py

```diff
@@ -0,0 +1,171 @@
+"""EDI stock move request records"""
+
+from odoo import api, fields, models, _
+from odoo.exceptions import ValidationError
+import logging
+
+_logger = logging.getLogger(__name__)
+
+
+class StockMove(models.Model):
+    """Extend ``stock.move`` to include the EDI Stock Cancellation document"""
+
+    _inherit = "stock.move"
+
+    edi_stock_cancellation_doc_id = fields.Many2one(
+        "edi.document",
+        string="EDI Stock Cancellation Document",
+        required=False,
+        readonly=True,
+        index=True,
+    )
+
+
+class EdiDocument(models.Model):
+    """Extend ``edi.document`` to include stock move cancellation records"""
+
+    _inherit = "edi.document"
+
+    move_cancellation_ids = fields.One2many(
+        "edi.record.stock.cancellation",
+        "doc_id",
+        string="Cancelled Stock Moves",
+    )
+
+
+class EdiStockCancellationDocument(models.AbstractModel):
+    """Stock Cancellation document model
+
+    Stock Cancellation documents contain sets of moves
+    (grouped by sale & picking_type) which have been cancelled
+    since the previous EDI export. Note this report only makes sense
+    for outgoing picks linked to sales
+    """
+
+    _name = "edi.document.stock.cancellation"
+    _inherit = "edi.move.report.document"
+    _description = "Stock Cancellations"
+
+    _edi_move_report_via = "edi_stock_cancellation_doc_id"
+
+    @api.model
+    def move_report_domain(self, _doc):
+        """
+        Override to only include cancelled moves for picking types which have
+        u_enable_pick_cancellation_email_notifs enabled.
+        _edi_move_report_via functionality is kept to ensure moves are not reported on twice
+        """
+        StockPickingType = self.env["stock.picking.type"]
+        picking_types_to_consider = StockPickingType.search(
+            [("u_enable_pick_cancellation_email_notifs", "=", True)]
+        )
+        domain = [
+            ("state", "=", "cancel"),
+            (self._edi_move_report_via, "=", False),
+            ("picking_id.picking_type_id", "in", picking_types_to_consider.ids),
+        ]
+        return domain
+
+    @api.model
+    def move_report_list(self, doc, moves):
+        """
+        Inherit to group moves by sale & picking_type to result
+        in a cancellation record for each unique combination
+        """
+        res = super().move_report_list(doc, moves)
+        grouped_moves = []
+        for sales_order, moves_by_sale in res.groupby(lambda move: move.sale_line_id.order_id):
```

Comment:
Not all customers that have `edi_stock` have sale orders. Would this still work for those customers, or does this need moving into its own module/ a different module?

Reply:
Ah, I was thinking this in the morning but didn't think about the `sale_line_id` column not existing in those instances.

Ideally I don't want to have a bridge module for this, so might just add a check for the column existing - if that sounds OK?

Reply:
discussed w/ miquel, going to add `edi_sale_stock` module in place of what i did with the nasty module check